File: python/lombokalgoritma/graph/a_star.py
# LombokAlgoritma — A* search (SPEC §9.4)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

import heapq
import math
from collections.abc import Callable
from typing import NamedTuple

from ..errors import NegativeWeightError
from .types import Graph, assert_node, build_adj_list


class AStarResult(NamedTuple):
    """Path ``source … target`` and its cost (``[]``, ``inf`` when unreachable)."""

    path: list[int]
    cost: float


def _zero(_: int) -> float:
    return 0


def a_star(
    g: Graph,
    source: int,
    target: int,
    heuristic: Callable[[int], float] | None = None,
) -> AStarResult:
    """A* with an admissible ``heuristic`` (default 0); heap keyed ``(g + h, node)``.

    Raises:
        NegativeWeightError: any edge weight is negative.
    """
    h = heuristic if heuristic is not None else _zero
    adj = build_adj_list(g)
    assert_node(g, source, "source")
    assert_node(g, target, "target")
    if any(w < 0 for _, _, w in g.edges):
        raise NegativeWeightError("a_star")
    g_score: list[float] = [math.inf] * len(adj)
    prev = [-1] * len(adj)
    g_score[source] = 0
    heap: list[tuple[float, int]] = [(h(source), source)]
    while heap:
        f, u = heapq.heappop(heap)
        gu = g_score[u]
        if f > gu + h(u):
            continue
        if u == target:
            path = [target]
            cur = prev[target]
            while cur != -1:
                path.append(cur)
                cur = prev[cur]
            path.reverse()
            return AStarResult(path, gu)
        for v, w in adj[u]:
            ng = gu + w
            if ng < g_score[v]:
                g_score[v] = ng
                prev[v] = u
                heapq.heappush(heap, (ng + h(v), v))
    return AStarResult([], math.inf)

File: typescript/src/graph/a-star.ts
// LombokAlgoritma — A* search
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import { NegativeWeightError } from '../core/errors.js';
import { MinHeap, tupleLess } from '../core/heap.js';
import { type Graph, assertNode, buildAdjList } from './types.js';

/**
 * A* from `source` to `target` with an admissible `heuristic` (default 0 ⇒ Dijkstra).
 * Frontier keyed by `[g + h, node]`; stale entries are skipped; a node's predecessor is updated
 * only on a strict improvement. Returns the path (source … target) and its cost, or
 * `{ path: [], cost: Infinity }` when unreachable. SPEC §6.5.
 *
 * @throws RangeError on a negative edge weight
 */
export function aStar(
  g: Graph,
  source: number,
  target: number,
  heuristic: (node: number) => number = () => 0,
): { path: number[]; cost: number } {
  const adj = buildAdjList(g);
  assertNode(g, source, 'source');
  assertNode(g, target, 'target');
  if (g.edges.some((e) => e.weight < 0)) throw new NegativeWeightError('aStar');
  const gScore = new Array<number>(g.nodes).fill(Number.POSITIVE_INFINITY);
  const prev = new Array<number>(g.nodes).fill(-1);
  gScore[source] = 0;
  const heap = new MinHeap<[number, number]>(tupleLess);
  heap.push([heuristic(source), source]);
  for (let top = heap.pop(); top !== undefined; top = heap.pop()) {
    const [f, u] = top;
    const gu = gScore[u] as number;
    if (f > gu + heuristic(u)) continue; // stale entry
    if (u === target) {
      const path = [target];
      for (let cur = prev[target] as number; cur !== -1; cur = prev[cur] as number) path.push(cur);
      return { path: path.reverse(), cost: gu };
    }
    for (const { to, weight } of adj[u] ?? []) {
      const ng = gu + weight;
      if (ng < (gScore[to] as number)) {
        gScore[to] = ng;
        prev[to] = u;
        heap.push([ng + heuristic(to), to]);
      }
    }
  }
  return { path: [], cost: Number.POSITIVE_INFINITY };
}
